src/controller_examples/rabbitmq_controller/proof/liveness/proof.rs

```diff
@@ -0,0 +1,319 @@
+    lemma_from_init_step_to_after_create_headless_service_step(spec, rabbitmq);
+
+    use_tla_forall(spec, |res: SubResource| always(lift_state(helper_invariants::every_resource_create_request_implies_at_after_create_resource_step(res, rabbitmq))), SubResource::ServerConfigMap);
+    assert forall |sub_resource: SubResource| sub_resource != SubResource::StatefulSet implies spec.entails(
+        lift_state(#[trigger] pending_req_in_flight_at_after_get_resource_step(sub_resource, rabbitmq))
+            .leads_to(lift_state(pending_req_in_flight_at_after_get_resource_step(next_resource_get_step_and_request(rabbitmq, sub_resource).0.get_AfterKRequestStep_1(), rabbitmq)))) by {
+        use_tla_forall_for_sub_resource(spec, sub_resource, rabbitmq);
+        let next_resource = next_resource_get_step_and_request(rabbitmq, sub_resource).0.get_AfterKRequestStep_1();
+        lemma_from_after_get_resource_step_to_resource_matches(spec, rabbitmq, sub_resource, next_resource);
+    }
+    // Thanks to the recursive construction of macro.
+    leads_to_trans_n!(
+        spec, true_pred(), lift_state(|s: RMQCluster| { !s.ongoing_reconciles().contains_key(rabbitmq.object_ref()) }),
+        lift_state(|s: RMQCluster| { !s.ongoing_reconciles().contains_key(rabbitmq.object_ref()) && s.scheduled_reconciles().contains_key(rabbitmq.object_ref())}),
+        lift_state(no_pending_req_at_rabbitmq_step_with_rabbitmq(rabbitmq, RabbitmqReconcileStep::Init)),
+        lift_state(pending_req_in_flight_at_after_get_resource_step(SubResource::HeadlessService, rabbitmq)),
+        lift_state(pending_req_in_flight_at_after_get_resource_step(SubResource::Service, rabbitmq)),
+        lift_state(pending_req_in_flight_at_after_get_resource_step(SubResource::ErlangCookieSecret, rabbitmq)),
+        lift_state(pending_req_in_flight_at_after_get_resource_step(SubResource::DefaultUserSecret, rabbitmq)),
+        lift_state(pending_req_in_flight_at_after_get_resource_step(SubResource::PluginsConfigMap, rabbitmq)),
+        lift_state(pending_req_in_flight_at_after_get_resource_step(SubResource::ServerConfigMap, rabbitmq)),
+        lift_state(pending_req_in_flight_at_after_get_resource_step(SubResource::ServiceAccount, rabbitmq)),
+        lift_state(pending_req_in_flight_at_after_get_resource_step(SubResource::Role, rabbitmq)),
+        lift_state(pending_req_in_flight_at_after_get_resource_step(SubResource::RoleBinding, rabbitmq)),
+        lift_state(pending_req_in_flight_at_after_get_resource_step(SubResource::StatefulSet, rabbitmq))
+    );
+    assert forall |sub_resource: SubResource| spec.entails(true_pred().leads_to(lift_state(#[trigger] sub_resource_state_matches(sub_resource, rabbitmq)))) by {
+        use_tla_forall_for_sub_resource(spec, sub_resource, rabbitmq);
+        let next_resource = if sub_resource == SubResource::StatefulSet { sub_resource } else {
```

Comment:
Can you document a bit more about the special handling of stateful set here?
I understand that in `lemma_true_leads_to_always_current_state_matches` you skip stateful set in the forall coz it requires special reasoning, but don't really understand the special handling here.

Reply:
Here it doesn't because the stateful set requires special reasoning. In Line 141, I skip stateful set because that `forall` is used to show that `at_after_get_step(sub_resource) ~> at_after_get_step(next_resource)`. But for stateful set, there is no `next_resource`, it's already the last one.